twib/twibd/TCPFrontend.cpp

```diff
@@ -0,0 +1,261 @@
+		max_fd = std::max(max_fd, fd);
+		FD_SET(fd, &readfds);
+
+		for (auto &c : clients) {
+			FD_SET(c->fd, &readfds);
+			max_fd = std::max(max_fd, c->fd);
+			if (c->out_buffer.size() > 0) {
+				FD_SET(c->fd, &writefds);
+			}
+		}
+
+		/*if (max_fd > FD_SETSIZE) {
+			log(FATAL, "FD is higher than set size: %d %d", max_fd, FD_SETSIZE);
+			exit(1);
+		}*/
+		if (select(max_fd + 1, &readfds, &writefds, NULL, NULL) < 0) {
+			log(FATAL, "failed to select file descriptors: %s", NET_ERR_STR);
+			exit(1);
+		}
+
+		// Check select status on server socket
+		if (FD_ISSET(fd, &readfds)) {
+			log(DEBUG, "incoming connection detected on %d", fd);
+
+			struct sockaddr_in6 client_addr;
+			socklen_t client_addrlen = sizeof(client_addr);
+			int client_fd = accept(fd, (struct sockaddr*)&client_addr, &client_addrlen);
+			if (client_fd < 0) {
+				log(WARN, "failed to accept incoming connection");
+			} else {
+				std::shared_ptr<Client> client = std::make_shared<Client>(this, client_fd);
+				clients.push_back(client);
+				twibd->AddClient(client);
+			}
+		}
+
+		// pump i/o
+		for (auto ci = clients.begin(); ci != clients.end(); ci++) {
+			std::shared_ptr<Client> &client = *ci;
+			if (FD_ISSET(client->fd, &writefds)) {
+				client->PumpOutput();
+			}
+			if (FD_ISSET(client->fd, &readfds)) {
+				log(DEBUG, "incoming data for client %x", client->client_id);
+				client->PumpInput();
+			}
+		}
+
+		for (auto i = clients.begin(); i != clients.end(); ) {
+			if ((*i)->deletion_flag) {
+				twibd->RemoveClient(*i);
+				i = clients.erase(i);
+				continue;
+			}
+
+			(*i)->Process();
+			i++;
+		}
+	}
+}
+
+void TCPFrontend::NotifyEventThread() {
+	// Connect to the server in order to wake it up
+	int fd = socket(AF_INET, SOCK_STREAM, 0);
+	if (fd < 0) {
+		log(FATAL, "failed to create socket: %s", NET_ERR_STR);
+		exit(1);
+	}
+	struct sockaddr_in addr;
+	memset(&addr, 0, sizeof(addr));
+	addr.sin_family = AF_INET;
+	addr.sin_port = htons(15151);
+	addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+
+	if (connect(fd, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
+		log(FATAL, "failed to connect for notification: %s", NET_ERR_STR);
+		exit(1);
+	}
+	close(fd);
+}
+
+TCPFrontend::Client::Client(TCPFrontend *frontend, int fd) : frontend(frontend), twibd(frontend->twibd), fd(fd) {
+}
+
+TCPFrontend::Client::~Client() {
+	log(DEBUG, "Closing connection %x", this->client_id);
+	close(fd);
+}
+
+void TCPFrontend::Client::PumpOutput() {
+	log(DEBUG, "pumping out 0x%lx bytes", out_buffer.size());
+	std::lock_guard<std::mutex> lock(out_buffer_mutex);
+	if (out_buffer.size() > 0) {
+		ssize_t r = send(fd, out_buffer.data(), out_buffer.size(), 0);
+		log(DEBUG, "Managed to send 0x%lx bytes", r);
+		if (r <= 0) {
+			deletion_flag = true;
+			return;
+		}
+		if (r > 0) {
+			// move everything that we didn't send to the start of the buffer
+			std::move(out_buffer.begin() + r, out_buffer.end(), out_buffer.begin());
+			out_buffer.resize(out_buffer.size() - r);
```

Comment:
ring buffers?

Reply:
Yeah I didn't move the code to ring buffers when I rebased (The PR was originally made before ring buffers made it in).